api/src/main/java/org/apache/iceberg/expressions/BoundReference.java

```diff
@@ -19,57 +19,238 @@
   }
 
   public T get(StructLike struct) {
-    return struct.get(pos, javaType());
+    return (T) accessor.get(struct);
   }
 
   @Override
   public String toString() {
-    return String.format("ref(id=%d, pos=%d, type=%s)", fieldId, pos, type);
+    return String.format("ref(id=%d, accessor=%s)", fieldId, accessor);
+  }
+
+  private interface Accessor<T> extends Serializable {
+    Object get(T container);
+
+    Type type();
+  }
+
+  private static class PositionAccessor implements Accessor<StructLike> {
+    private int p;
+    private final Type type;
+    private final Class<?> javaClass;
+
+    private PositionAccessor(int p, Type type) {
+      this.p = p;
+      this.type = type;
+      this.javaClass = type.typeId().javaClass();
+    }
+
+    @Override
+    public Object get(StructLike row) {
+      return row.get(p, javaClass);
+    }
+
+    @Override
+    public Type type() {
+      return type;
+    }
+  }
+
+  private static class Position2Accessor implements Accessor<StructLike> {
+    private final int p0;
+    private final int p1;
+    private final Type type;
+    private final Class<?> javaClass;
+
+    private Position2Accessor(int p, PositionAccessor wrapped) {
+      this.p0 = p;
+      this.p1 = wrapped.p;
+      this.type = wrapped.type;
+      this.javaClass = wrapped.javaClass;
+    }
+
+    @Override
+    public Object get(StructLike row) {
+      return row.get(p0, StructLike.class).get(p1, javaClass);
+    }
+
+    @Override
+    public Type type() {
+      return type;
+    }
+  }
+
+  private static class Position3Accessor implements Accessor<StructLike> {
+    private final int p0;
+    private final int p1;
+    private final int p2;
+    private final Type type;
+    private final Class<?> javaClass;
+
+    private Position3Accessor(int p, Position2Accessor wrapped) {
+      this.p0 = p;
+      this.p1 = wrapped.p0;
+      this.p2 = wrapped.p1;
+      this.type = wrapped.type;
+      this.javaClass = wrapped.javaClass;
+    }
+
+    @Override
+    public Object get(StructLike row) {
+      return row.get(p0, StructLike.class).get(p1, StructLike.class).get(p2, javaClass);
+    }
+
+    @Override
+    public Type type() {
+      return type;
+    }
+  }
+
+  private static class WrappedPositionAccessor implements Accessor<StructLike> {
+    private final int p;
+    private final Accessor<StructLike> accessor;
+
+    private WrappedPositionAccessor(int p, Accessor<StructLike> accessor) {
+      this.p = p;
+      this.accessor = accessor;
+    }
+
+    @Override
+    public Object get(StructLike row) {
+      StructLike inner = row.get(p, StructLike.class);
+      if (inner != null) {
+        return accessor.get(inner);
+      }
+      return null;
+    }
+
+    @Override
+    public Type type() {
+      return accessor.type();
+    }
+  }
+
+  private transient Map<Integer, Accessor<StructLike>> idToAccessor = null;
+
+  private Map<Integer, Accessor<StructLike>> lazyIdToAccessor(Schema schema) {
```

Comment:
The lazy pattern doesn't make sense here where a `BoundReference` is for just one predicate. There's no need to get accessors for the whole schema or keep the result around.

Maybe accessors should be provided by the `Schema`? Or maybe this should be in the `Binder`. That would keep `Schema` clean and reuse the same accessors for the schema.

Reply:
So i tried putting this into `Binder` but that is not the only place predicates are  bound seems like.  `UnboundPredicate.bind()` is called from `Projections`, `ResidualEvaluator`,  and `Binder` .  I also tried putting it in `UnboundPredicate` but that is not useful coz there's one created for every predicate. 

Seems like Schema  is a good place? So i'm putting it there. Lemme know if this looks any better to you.

Reply:
Side effect of this is `Accessor`  interface had to be made public